src/main/webapp/app/guided-tour/guided-tour.service.ts

```diff
@@ -0,0 +1,396 @@
+         */
+        fromEvent(window, 'resize')
+            .pipe(debounceTime(200))
+            .subscribe(() => {
+                if (this.currentTour && this.currentTourStepIndex > -1) {
+                    if (this.currentTour.minimumScreenSize && this.currentTour.minimumScreenSize >= window.innerWidth) {
+                        this.onResizeMessage = true;
+                        this.guidedTourCurrentStepSubject.next({
+                            headlineTranslateKey: 'tour.resize.headline',
+                            contentType: ContentType.TEXT,
+                            contentTranslateKey: 'tour.resize.content',
+                        });
+                    } else {
+                        this.onResizeMessage = false;
+                        this.guidedTourCurrentStepSubject.next(this.getPreparedTourStep(this.currentTourStepIndex));
+                    }
+                }
+            });
+    }
+
+    /**
+     * @return current tour step as Observable
+     */
+    public getGuidedTourCurrentStepStream(): Observable<TourStep | null> {
+        return this.guidedTourCurrentStepSubject.asObservable();
+    }
+
+    /**
+     * Load course overview tour
+     * @return guided tour `courseOverviewTour`
+     */
+    public getOverviewTour(): Observable<GuidedTour> {
+        return of(courseOverviewTour);
+    }
+
+    /**
+     * Check if the provided tour step is the currently active one
+     * @param tourStep: current tour step of the guided tour
+     */
+    public isCurrentStep(tourStep: TourStep): boolean {
+        if (this.currentTour && this.currentTour.steps) {
+            return this.currentTourStepDisplay === this.currentTour.steps.indexOf(tourStep) + 1;
+        }
+        return false;
+    }
+
+    /**
+     * Navigate to next tour step
+     */
+    public nextStep(): void {
+        if (!this.currentTour) {
+            return;
+        }
+        const currentStep = this.currentTour.steps[this.currentTourStepIndex];
+        if (currentStep.closeAction) {
+            currentStep.closeAction();
+        }
+        if (this.currentTour.steps[this.currentTourStepIndex + 1]) {
+            this.currentTourStepIndex++;
+            if (currentStep.action) {
+                currentStep.action();
+            }
+            // Usually an action is opening something so we need to give it time to render.
+            setTimeout(() => {
+                if (this.checkSelectorValidity()) {
+                    this.guidedTourCurrentStepSubject.next(this.getPreparedTourStep(this.currentTourStepIndex));
+                } else {
+                    this.nextStep();
+                }
+            });
+        } else {
+            if (this.currentTour.completeCallback) {
+                this.currentTour.completeCallback();
+            }
+            this.updateGuidedTourSettings(this.currentTour.settingsKey, this.currentTourStepDisplay).subscribe(guidedTourSettings => {
```

Comment:
Does this mean that you save the guided tour settings after each step? I am not sure if this is a good idea, because it could involve a lot of REST calls on the server. Can you save the guided tour setting on the server only if the user closes the guided tour (e.g. by pressing Escape)?

If you want to make sure 100% that intermediate steps are not forgotten, you can always save them in the local browser storage

Reply:
@krusche The update is only done when the user finishes the guided tour or closing it by clicking the close icon or pressing escape. So that is already implemented :)

Reply:
Are you sure? Why do you call the updateGuidedTourSettings(...) then in nextStep()
It is not obvious what you do here. The code is not self-explanatory!

I suggest that you 
1) refactor the code a bit because the method is already 30 : extract a method "finishTour()" that is invoked appropriately and that calls updateGuidedTourSettings(...)
2) Add more comments so that other developers understand the code more easily!

Thanks